### PR TITLE
opencv3: fix for long broken linuxbrew (3 unrelated errors)

### PR DESCRIPTION
**Pull request by timotheecour**
@sjackman  @jonchang

- [x] Followed the guidelines in our [Contributing](https://github.com/Homebrew/homebrew-science/blob/master/.github/CONTRIBUTING.md) document?
- [x] Checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/homebrew-science/pulls) for the same update/change?
- [x] Run `brew audit --strict --online <formula>` (where `<formula>` is the name of the formula you're submitting) and corrected all errors?
- [x] Built your formula locally prior to submission with `brew install <formula>`?

fixes https://github.com/Homebrew/homebrew-science/issues/4825 which was just the tip of the iceberg...

### CODE REVIEW

**Comment by sjackman** [comment]
It built bottles for Linuxbrew on CircleCI. That's good. `brew test` failed:
```
==> brew test homebrew/science/opencv3 --verbose
Testing homebrew/science/opencv3
==> /usr/bin/g++-4.8 test.cpp -I/home/linuxbrew/.linuxbrew/Cellar/opencv3/3.2.0/include -L/home/linuxbrew/.linuxbrew/Cellar/opencv3/3.2.0/lib -o test

==> python -c 'import cv2; print(cv2.__version__)'
Illegal instruction (core dumped)
Error: homebrew/science/opencv3: failed
<0> expected but was
<132>.
==> FAILED
```
`Illegal instruction` usually indicates that an instruction was used that is not available on this machine, for example using an SSE 4.2 instruction on a machine that does not have them. If you have access to a Linux machine, you can test the bottle out on that machine:
https://circleci.com/gh/Homebrew/homebrew-science/3596#artifacts/containers/0
https://3596-4106792-gh.circle-artifacts.com/0/tmp/circle-artifacts.7q95bYj/opencv3-3.2.0.x86_64_linux.bottle.1.tar.gz

**Comment by timotheecour** [comment]
```
brew2 test homebrew/science/opencv3
Error: Testing requires the latest version of homebrew/science/opencv3
```

`brew2 help test` doesn't show a way to test from a file; what command do u want me to run? and I guess docs should be updated for that

File: opencv3.rb [hunk 1]
```diff
@@ -107,13 +107,24 @@ def install
...
+
+    # cf https://github.com/Homebrew/homebrew-science/pull/5185
+    args << "-DBUILD_OPENEXR=" + (OS.linux? ? "ON" : "OFF")
+
+    if OS.linux?
+      # http://answers.opencv.org/question/121651/fata-error-lapacke_h_path-notfound-when-building-opencv-32/
+      args << "-DWITH_LAPACK=OFF"
+
+      # avoid error: '_mm_cvtps_ph' was not declared in this scope; cf https://github.com/RoboSherlock/robosherlock/issues/78#issuecomment-274469830
+      args << "-DOPENCV_EXTRA_CXX_FLAGS='-march=core-avx2'"
```
Comment (sjackman) [inline opencv3.rb]: This line looks good, assuming that there's no way to configure opencv3 not to use AVX2.
Note to me: `-march=core-avx2` is a synonym for `haswell`.
I found these useful resources for `-march` flags:
https://gcc.gnu.org/onlinedocs/gcc-6.3.0/gcc/x86-Options.html#x86-Options
https://github.com/gcc-mirror/gcc/blob/master/gcc/config/i386/driver-i386.c#L772
https://raw.githubusercontent.com/gcc-mirror/gcc/master/gcc/config/i386/i386.c

File: opencv3.rb [hunk 2]
```diff
@@ -107,13 +107,24 @@ def install
...
-      -DBUILD_OPENEXR=OFF
       -DBUILD_PNG=OFF
       -DBUILD_ZLIB=OFF
       -DCMAKE_OSX_DEPLOYMENT_TARGET=
       -DJPEG_INCLUDE_DIR=#{jpeg.opt_include}
       -DJPEG_LIBRARY=#{jpeg.opt_lib}/libjpeg.#{dylib}
     ]
+
+    # cf https://github.com/Homebrew/homebrew-science/pull/5185
+    args << "-DBUILD_OPENEXR=" + (OS.linux? ? "ON" : "OFF")
```
Comment (sjackman) [inline opencv3.rb]: Note to self: this line was merged for `opencv` in #5185.

**Comment by sjackman** [comment]
> could u suggest another way? still, it's an improvement over the current status where nothing builds at all; so maybe we could merge this PR in the meantime and keep the bug open until the test is fixed

Two options:
1. Is it possible to configure `opencv3` not to use AVX2?
2. Modify brew test to check whether the CPU supports AVX2 and return if it does not.

```ruby
test do
  unless Hardware::CPU::avx2?
    owarn "brew test opencv3 requires AVX2"
    return
   end
  …
end

**Comment by iMichka** [comment]
I would also be in favour to merge this as soon as possible and open a new issue for the failing test, so we have more time to look at it later.

**Comment by sjackman** [comment]
Merged without a bottle because of the failed test. I think this formula needs two things:
1. `depends_on "openblas" unless OS.mac?` and remove `-DWITH_LAPACK=OFF`
2. Skip the test `unless Hardware::CPU::avx2?` (see https://github.com/Homebrew/homebrew-science/pull/5186#issuecomment-283423918)

**Comment by iMichka** [comment]
Closed by d829670a5d2647338fcbc2384017ac2341cab2f8

I'll have a look at the missing things to get bottles built.

**Pull request closed without merging**